strawberryfields/engine.py

```diff
@@ -127,870 +78,173 @@
 
-        To avoid discrepancies with the backend this method must not be called directly,
-        since it bypasses the circuit queue and the associated RegRef bookkeeping.
+        This method should not be called directly.
 
         Args:
-            clist (list[Command]): command list to run
-
+            prog (Program): program to run
         Returns:
             list[Command]: commands that were applied to the backend
         """
         applied = []
-        for cmd in clist:
-            if cmd.op is None:
-                # None represents an identity gate
-                continue
-            else:
+        for cmd in prog.circuit:
+            try:
+                # try to apply it to the backend
+                cmd.op.apply(cmd.reg, self.backend, hbar=self.hbar, **kwargs)
+                applied.append(cmd)
+            except NotApplicableError:
+                # command is not applicable to the current backend type
+                raise NotApplicableError('The operation {} cannot be used with {}.'.format(cmd.op, self.backend)) from None
+            except NotImplementedError:
+                # command not directly supported by backend API, try a decomposition instead
                 try:
-                    # try to apply it to the backend
-                    cmd.op.apply(cmd.reg, self.backend,
-                                 hbar=self.hbar, **kwargs)
-                    applied.append(cmd)
-                except NotApplicableError:
-                    # command is not applicable to the current backend type
-                    raise NotApplicableError('The operation {} cannot be used with {}.'.format(
-                        cmd.op, self.backend)) from None
-                except NotImplementedError:
-                    # command not directly supported by backend API, try a decomposition instead
-                    try:
-                        temp = cmd.op.decompose(cmd.reg)
-                        # run the decomposition
-                        applied_cmds = self._run_command_list(temp)
-                        # todo should we store the decomposition or the original Command?
-                        # if we change backends and re-run the circuit, the decomposition
-                        # may not be valid or useful.
-                        applied.extend(applied_cmds)
-                    except NotImplementedError as err:
-                        # simplify the error message by suppressing the previous exception
-                        raise err from None
+                    temp = cmd.op.decompose(cmd.reg)
+                    # run the decomposition
+                    applied_cmds = self._run_command_list(temp)
+                    applied.extend(applied_cmds)
+                except NotImplementedError as err:
+                    # simplify the error message by suppressing the previous exception
+                    raise err from None
```

Comment:
This logic for applying quantum operations seems by and large unchanged - how come it broke the `Decomposition` subclass of operations?

Reply:
It's the hbar handling which is the issue with `ops.Decomposition`